File: frontend/src/types/usage.ts
/**
 * Usage Metering Types
 * Types for usage tracking, quotas, and alerts
 */

// ============================================================================
// METRIC TYPES
// ============================================================================

export type MetricType = 'api_calls' | 'storage_bytes' | 'active_users' | 'branches';

export const METRIC_DISPLAY_NAMES: Record<MetricType, string> = {
  api_calls: 'API Calls',
  storage_bytes: 'Storage',
  active_users: 'Active Users',
  branches: 'Branches',
};

export const METRIC_UNITS: Record<MetricType, string> = {
  api_calls: 'calls',
  storage_bytes: 'bytes',
  active_users: 'users',
  branches: 'branches',
};

// ============================================================================
// USAGE QUOTA
// ============================================================================

export interface UsageQuota {
  id: string;
  tenant_id: string;
  metric_type: MetricType;
  limit_value: number;
  current_value: number;
  period_start: string;
  reset_date: string | null;
  alert_threshold: number;
  usage_percentage: number;
  is_unlimited: boolean;
  is_exceeded: boolean;
  is_near_limit: boolean;
  remaining: number;
  created_at: string;
  updated_at: string | null;
}

export interface UsageQuotaUpdate {
  limit_value?: number;
  alert_threshold?: number;
  reset_date?: string;
}

export interface UsageQuotaListResponse {
  items: UsageQuota[];
  total: number;
}

// ============================================================================
// USAGE ALERT
// ============================================================================

export type AlertType = 'threshold_warning' | 'limit_reached' | 'limit_exceeded';

export interface UsageAlert {
  id: string;
  tenant_id: string;
  metric_type: MetricType;
  alert_type: AlertType;
  usage_percentage: number;
  current_value: number;
  limit_value: number;
  message: string | null;
  acknowledged_at: string | null;
  created_at: string;
}

export interface UsageAlertListResponse {
  items: UsageAlert[];
  total: number;
}

// ============================================================================
// USAGE SUMMARY
// ============================================================================

export interface UsageMetricSummary {
  metric_type: MetricType;
  metric_display_name: string;
  current_value: number;
  limit_value: number;
  usage_percentage: number;
  is_unlimited: boolean;
  is_exceeded: boolean;
  is_near_limit: boolean;
  remaining: number;
  unit: string;
}

export interface TenantUsageSummary {
  tenant_id: string;
  tenant_name: string | null;
  period_start: string;
  metrics: UsageMetricSummary[];
  has_alerts: boolean;
  unacknowledged_alerts: number;
}

// ============================================================================
// USAGE TRENDS
// ============================================================================

export interface UsageTrendPoint {
  date: string;
  value: number;
}

export interface UsageTrends {
  metric_type: MetricType;
  start_date: string;
  end_date: string;
  data_points: UsageTrendPoint[];
  total: number;
  average: number;
}

// ============================================================================
// ADMIN OVERVIEW
// ============================================================================

export interface TenantUsageOverview {
  tenant_id: string;
  tenant_name: string;
  tier: string;
  api_calls: number;
  api_calls_limit: number;
  api_calls_percentage: number;
  storage_bytes: number;
  storage_limit_bytes: number;
  storage_percentage: number;
  active_users: number;
  users_limit: number;
  users_percentage: number;
  branches: number;
  branches_limit: number;
  branches_percentage: number;
  has_exceeded: boolean;
  has_warning: boolean;
}

export interface AdminUsageOverviewResponse {
  items: TenantUsageOverview[];
  total: number;
  tenants_with_warnings: number;
  tenants_exceeded: number;
}

// ============================================================================
// USAGE INCREMENT
// ============================================================================

export interface UsageIncrementRequest {
  metric_type: MetricType;
  amount?: number;
}

export interface UsageIncrementResponse {
  metric_type: MetricType;
  previous_value: number;
  new_value: number;
  limit_value: number;
  usage_percentage: number;
  is_exceeded: boolean;
  is_near_limit: boolean;
  remaining: number;
}

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

export function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  if (bytes < 1024 * 1024 * 1024) return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
  return `${(bytes / (1024 * 1024 * 1024)).toFixed(2)} GB`;
}

export function formatMetricValue(metricType: MetricType, value: number): string {
  if (metricType === 'storage_bytes') {
    return formatBytes(value);
  }
  if (value === -1) return 'Unlimited';
  return value.toLocaleString();
}

export function formatLimit(metricType: MetricType, value: number): string {
  if (value === -1) return 'Unlimited';
  if (metricType === 'storage_bytes') {
    return formatBytes(value);
  }
  return value.toLocaleString();
}

export function getUsagePercentageColor(percentage: number, threshold: number = 80): string {
  if (percentage >= 100) return 'text-red-600 dark:text-red-400';
  if (percentage >= threshold) return 'text-amber-600 dark:text-amber-400';
  if (percentage >= 50) return 'text-blue-600 dark:text-blue-400';
  return 'text-emerald-600 dark:text-emerald-400';
}

export function getUsageProgressColor(percentage: number, threshold: number = 80): string {
  if (percentage >= 100) return 'bg-red-500';
  if (percentage >= threshold) return 'bg-amber-500';
  if (percentage >= 50) return 'bg-blue-500';
  return 'bg-emerald-500';
}

export function getAlertTypeColor(alertType: AlertType): string {
  switch (alertType) {
    case 'limit_exceeded':
      return 'bg-red-100 text-red-800 dark:bg-red-900/30 dark:text-red-400';
    case 'limit_reached':
      return 'bg-amber-100 text-amber-800 dark:bg-amber-900/30 dark:text-amber-400';
    case 'threshold_warning':
      return 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900/30 dark:text-yellow-400';
    default:
      return 'bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-300';
  }
}

export function getAlertTypeLabel(alertType: AlertType): string {
  switch (alertType) {
    case 'limit_exceeded':
      return 'Limit Exceeded';
    case 'limit_reached':
      return 'Limit Reached';
    case 'threshold_warning':
      return 'Warning';
    default:
      return alertType;
  }
}

File: backend/app/schemas/usage.py
"""
Usage Schemas
Request/response models for usage metering and quotas
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Literal
from datetime import datetime, date
from uuid import UUID


# ============================================================================
# METRIC TYPES
# ============================================================================

MetricTypeEnum = Literal["api_calls", "storage_bytes", "active_users", "branches"]


# ============================================================================
# USAGE RECORD SCHEMAS
# ============================================================================

class UsageRecordCreate(BaseModel):
    """Schema for creating a usage record"""
    tenant_id: UUID
    metric_type: MetricTypeEnum
    value: int = Field(..., ge=0)
    recorded_date: Optional[date] = None


class UsageRecordResponse(BaseModel):
    """Usage record response"""
    id: UUID
    tenant_id: UUID
    metric_type: str
    value: int
    recorded_date: date
    created_at: datetime

    class Config:
        from_attributes = True


class UsageRecordListResponse(BaseModel):
    """List of usage records"""
    items: List[UsageRecordResponse]
    total: int


# ============================================================================
# USAGE QUOTA SCHEMAS
# ============================================================================

class UsageQuotaCreate(BaseModel):
    """Schema for creating a usage quota"""
    metric_type: MetricTypeEnum
    limit_value: int = Field(..., ge=-1, description="-1 for unlimited")
    alert_threshold: int = Field(default=80, ge=0, le=100)


class UsageQuotaUpdate(BaseModel):
    """Schema for updating a usage quota"""
    limit_value: Optional[int] = Field(None, ge=-1)
    alert_threshold: Optional[int] = Field(None, ge=0, le=100)
    reset_date: Optional[date] = None


class UsageQuotaResponse(BaseModel):
    """Usage quota response"""
    id: UUID
    tenant_id: UUID
    metric_type: str
    limit_value: int
    current_value: int
    period_start: date
    reset_date: Optional[date]
    alert_threshold: int
    usage_percentage: float
    is_unlimited: bool
    is_exceeded: bool
    is_near_limit: bool
    remaining: int
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


class UsageQuotaListResponse(BaseModel):
    """List of usage quotas"""
    items: List[UsageQuotaResponse]
    total: int


# ============================================================================
# USAGE ALERT SCHEMAS
# ============================================================================

class UsageAlertResponse(BaseModel):
    """Usage alert response"""
    id: UUID
    tenant_id: UUID
    metric_type: str
    alert_type: str
    usage_percentage: int
    current_value: int
    limit_value: int
    message: Optional[str]
    acknowledged_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


class UsageAlertListResponse(BaseModel):
    """List of usage alerts"""
    items: List[UsageAlertResponse]
    total: int


class UsageAlertAcknowledge(BaseModel):
    """Schema for acknowledging an alert"""
    alert_id: UUID


# ============================================================================
# USAGE SUMMARY SCHEMAS
# ============================================================================

class UsageMetricSummary(BaseModel):
    """Summary of a single metric"""
    metric_type: str
    metric_display_name: str
    current_value: int
    limit_value: int
    usage_percentage: float
    is_unlimited: bool
    is_exceeded: bool
    is_near_limit: bool
    remaining: int
    unit: str  # "calls", "bytes", "users", etc.


class TenantUsageSummary(BaseModel):
    """Complete usage summary for a tenant"""
    tenant_id: UUID
    tenant_name: Optional[str]
    period_start: date
    metrics: List[UsageMetricSummary]
    has_alerts: bool
    unacknowledged_alerts: int


class UsageTrendPoint(BaseModel):
    """Single data point for usage trends"""
    date: date
    value: int


class UsageTrends(BaseModel):
    """Time-series usage data"""
    metric_type: str
    start_date: date
    end_date: date
    data_points: List[UsageTrendPoint]
    total: int
    average: float


# ============================================================================
# ADMIN USAGE OVERVIEW
# ============================================================================

class TenantUsageOverview(BaseModel):
    """Usage overview for a single tenant (admin view)"""
    tenant_id: UUID
    tenant_name: str
    tier: str
    api_calls: int
    api_calls_limit: int
    api_calls_percentage: float
    storage_bytes: int
    storage_limit_bytes: int
    storage_percentage: float
    active_users: int
    users_limit: int
    users_percentage: float
    branches: int
    branches_limit: int
    branches_percentage: float
    has_exceeded: bool
    has_warning: bool


class AdminUsageOverviewResponse(BaseModel):
    """Admin usage overview response"""
    items: List[TenantUsageOverview]
    total: int
    tenants_with_warnings: int
    tenants_exceeded: int


# ============================================================================
# USAGE INCREMENT/DECREMENT
# ============================================================================

class UsageIncrementRequest(BaseModel):
    """Request to increment usage"""
    metric_type: MetricTypeEnum
    amount: int = Field(default=1, ge=1)


class UsageIncrementResponse(BaseModel):
    """Response after incrementing usage"""
    metric_type: str
    previous_value: int
    new_value: int
    limit_value: int
    usage_percentage: float
    is_exceeded: bool
    is_near_limit: bool
    remaining: int


class UsageResetRequest(BaseModel):
    """Request to reset usage counter"""
    metric_type: MetricTypeEnum


# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

METRIC_DISPLAY_NAMES = {
    "api_calls": "API Calls",
    "storage_bytes": "Storage",
    "active_users": "Active Users",
    "branches": "Branches",
}

METRIC_UNITS = {
    "api_calls": "calls",
    "storage_bytes": "bytes",
    "active_users": "users",
    "branches": "branches",
}


def format_bytes(bytes_value: int) -> str:
    """Format bytes to human-readable string"""
    if bytes_value < 1024:
        return f"{bytes_value} B"
    elif bytes_value < 1024 * 1024:
        return f"{bytes_value / 1024:.1f} KB"
    elif bytes_value < 1024 * 1024 * 1024:
        return f"{bytes_value / (1024 * 1024):.1f} MB"
    else:
        return f"{bytes_value / (1024 * 1024 * 1024):.2f} GB"


def format_metric_value(metric_type: str, value: int) -> str:
    """Format metric value based on type"""
    if metric_type == "storage_bytes":
        return format_bytes(value)
    elif value == -1:
        return "Unlimited"
    else:
        return f"{value:,}"
